File: Leetcode/python/sol733.py
# https://leetcode.com/problems/flood-fill/

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        S = [(sr, sc)]
        m = len(image[0])  # col
        n = len(image)  # rows
        original = image[sr][sc]
        seen = set()
        while S:
            row, col = S.pop(0)
            if image[row][col] == original:
                image[row][col] = color
            else:
                continue
            if (row, col) in seen:
                continue
            seen.add((row, col))
            # up
            if row - 1 >= 0:
                S.append((row - 1, col))
            # down
            if row + 1 < n:
                S.append((row + 1, col))
            # left
            if col - 1 >= 0:
                S.append((row, col - 1))
            # right
            if col + 1 < m:
                S.append((row, col + 1))

        return image

File: Leetcode/typescript/sol733.ts

// https://leetcode.com/problems/flood-fill/description/


function floodFill(image: number[][], sr: number, sc: number, color: number): number[][] {
    const directions = [[0,-1], [0,1], [1,0], [-1,0]]
    const stack = [[sr, sc]];
    const original = image[sr][sc];
    if(original === color){
        return image;
    }
    image[sr][sc] = color;
    while(0 < stack.length){
        const node = stack.pop();
        if(!node){
            continue;
        }
        const [row, col] = node;
        for(const [i, j] of directions){
            const next_row = row + i;
            const next_col = col + j;
            if ( 0 <= next_row &&
                next_row < image.length &&
                0 <= next_col &&
                next_col < image[0].length &&
                image[next_row][next_col] === original){
                    image[next_row][next_col] = color;
                    stack.push([next_row, next_col])
            }
        }
    }
    return image;
};